File: python/0141.py
# Link: https://leetcode.com/problems/linked-list-cycle/description/
# Runtime: 61 ms
# Memory: 20.31 MB

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        fast, slow = head, head
        hasCycle = False
        while fast != None and fast.next != None and not hasCycle:
            fast = fast.next.next
            slow = slow.next

            if fast == slow: hasCycle = True

        return hasCycle

File: c++/0141.cpp
// Link: https://leetcode.com/problems/linked-list-cycle/description/
// Runtime: 7 ms
// Memory: 8.11 MB

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
  bool hasCycle(ListNode *head) {
    ListNode *fast = head, *slow = head;
    bool hasCycle = false;
    while ((fast != nullptr) && (fast->next != nullptr) && (!hasCycle)) {
      fast = fast->next->next;
      slow = slow->next;

      if (fast == slow)
        hasCycle = true;
    }

    return hasCycle;
  }
};
